### PR TITLE
[Common optimizations][ReduceMerge] Align shapes and types before merging reduce nodes

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Details:

Shapes and types are aligned before merging reduce nodes.

### Tickets:
130581

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp [hunk 1]
```diff
@@ -64,6 +64,22 @@ bool fuse_reduce_operations(const std::shared_ptr<Node>& node) {
...
+        if (reduce_axes_rank == Dimension(0)) {
+            const auto unsqueeze_const =
+                ov::op::v0::Constant::create(reduce_axes_output.get_node()->get_element_type(), {}, {0});
+            const auto unsqueeze = std::make_shared<ov::op::v0::Unsqueeze>(reduce_axes_output, unsqueeze_const);
+            reduce->inputs()[1].replace_source_output(unsqueeze);
+            copy_runtime_info({top_reduce, bottom_reduce}, {unsqueeze_const, unsqueeze});
+        }
+        const auto cast = std::make_shared<ov::op::v0::Convert>(reduce->input_value(1), ov::element::i64);
+        reduce->inputs()[1].replace_source_output(cast);
+        copy_runtime_info({top_reduce, bottom_reduce}, cast);
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp:80]: I believe we need to copy rt_info from reduce axis node here, not from reduce op

File: src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp [hunk 2]
```diff
@@ -64,6 +64,22 @@ bool fuse_reduce_operations(const std::shared_ptr<Node>& node) {
...
+        const auto reduce_axes_output = reduce->input_value(1);
+        const auto reduce_axes_rank = reduce_axes_output.get_partial_shape().rank();
+        if (reduce_axes_rank == Dimension(0)) {
+            const auto unsqueeze_const =
+                ov::op::v0::Constant::create(reduce_axes_output.get_node()->get_element_type(), {}, {0});
+            const auto unsqueeze = std::make_shared<ov::op::v0::Unsqueeze>(reduce_axes_output, unsqueeze_const);
+            reduce->inputs()[1].replace_source_output(unsqueeze);
+            copy_runtime_info({top_reduce, bottom_reduce}, {unsqueeze_const, unsqueeze});
+        }
+        const auto cast = std::make_shared<ov::op::v0::Convert>(reduce->input_value(1), ov::element::i64);
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp:78]: It's better to check types first, and do this only if they are not equal

File: src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp [hunk 3]
```diff
@@ -64,6 +64,22 @@ bool fuse_reduce_operations(const std::shared_ptr<Node>& node) {
...
+        reduce->inputs()[1].replace_source_output(cast);
+        copy_runtime_info({top_reduce, bottom_reduce}, cast);
+    }
+
     std::shared_ptr<Node> axes =
         std::make_shared<ov::op::v0::Concat>(OutputVector{top_reduce->input_value(1), bottom_reduce->input_value(1)},
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp:88]: It should, but it is possible for reduce node to have axes as a scalar
image

File: src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp [hunk 4]
```diff
@@ -64,6 +64,22 @@ bool fuse_reduce_operations(const std::shared_ptr<Node>& node) {
...
+    // Align reduce axes constants by shape and type
+    for (auto& reduce : {top_reduce, bottom_reduce}) {
+        const auto reduce_axes_output = reduce->input_value(1);
+        const auto reduce_axes_rank = reduce_axes_output.get_partial_shape().rank();
+        if (reduce_axes_rank == Dimension(0)) {
+            const auto unsqueeze_const =
+                ov::op::v0::Constant::create(reduce_axes_output.get_node()->get_element_type(), {}, {0});
+            const auto unsqueeze = std::make_shared<ov::op::v0::Unsqueeze>(reduce_axes_output, unsqueeze_const);
+            reduce->inputs()[1].replace_source_output(unsqueeze);
+            copy_runtime_info({top_reduce, bottom_reduce}, {unsqueeze_const, unsqueeze});
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp]: Done

File: src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp (hunk 1 above)
Comment (daniil-lyakhov) [reply src/common/transformations/src/transformations/common_optimizations/reduce_merge.cpp]: Done

**Review by itikhono — APPROVED** [review]

**Pull request merged**